### DISCUSSION TITLE
Need help with atomFamily generic types

### DISCUSSION THREAD

**Discussion by oferitz** [pmndrs/jotai#2298] [Q&A] [answered]
I'm having difficulty grasping the proper typing of atomFamily with generics. Transitioning from Recoil, it was relatively straightforward when I needed an atomFamily with identical data structures for each atom value, but with one property capable of having a generic type. In such cases, I would define the atomFamily like this:
```ts
import { atomFamily as recoilAtomFamily } from "recoil";

type ComponentState<T = any> = {
  id: string;
  data: T;
};

type ExampleComponentState = {
    name: string;
    enabled: boolean;
 };

export const recoilComponentStateAtomFamily = recoilAtomFamily<
  ComponentState,
  string
>({
  key: "recoilComponentStateAtomFamily",
  default: (id: string) => {
    return getDefaultValue(id);
  },
});

//... and then using it in the component

  const [state, setState] = useRecoilState<
    ComponentState<ExampleComponentState>
  >(recoilComponentStateAtomFamily(id));

```

And this setup functions as expected, with both `state` and `setState` correctly typed.

Now the same with Jotai
```ts
import { atom, PrimitiveAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import deepEqual from "fast-deep-equal";

export const jotaiComponentStateAtomFamily = atomFamily<
  string,
  PrimitiveAtom<ComponentState>
>((id) => atom(getDefaultValue(id)), deepEqual);

//... and then using it in the component
  import { useAtom } from "jotai";
  
  const [state, setState] = useAtom<ComponentState<ExampleComponentState>>(
    jotaiComponentStateAtomFamily(id)
  );
  ```
  
However, the Jotai approach does not behave as anticipated. When attempting to use `setState`, TypeScript raises an error stating, "This expression is not callable. Type 'never' has no call signatures."
  
I've created a [Codesandbox](https://codesandbox.io/p/sandbox/recoil-vs-jotai-atomfamily-ts-yxscv3?file=%2Fsrc%2FRecoilComponent.tsx%3A2%2C55&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpztlz8t00063b6g45vomp1i%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpztlz8t00023b6gtelsn5m0%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpztlz8t00033b6g0zg6yjrt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpztlz8t00053b6gz7y9mv7a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B51.868647240631155%252C48.131352759368845%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpztlz8t00023b6gtelsn5m0%2522%253A%257B%2522id%2522%253A%2522clpztlz8t00023b6gtelsn5m0%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq0wi58w00bj3b6ghghiorgl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A1%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A122%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clq22hsq000023b6hs45snxuu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A2%252C%2522startColumn%2522%253A55%252C%2522endLineNumber%2522%253A2%252C%2522endColumn%2522%253A55%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FRecoilComponent.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clq22hsq000023b6hs45snxuu%2522%257D%252C%2522clpztlz8t00053b6gz7y9mv7a%2522%253A%257B%2522id%2522%253A%2522clpztlz8t00053b6gz7y9mv7a%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpztlz8t00043b6g0jvp1x9i%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpztlz8t00043b6g0jvp1x9i%2522%257D%252C%2522clpztlz8t00033b6g0zg6yjrt%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpztlz8t00033b6g0zg6yjrt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)  example that provides a comprehensive demonstration in both Recoil and Jotai. This should make it easier to grasp the complete context of what I'm attempting to achieve.

**Comment by dai-shi** [comment]
It's not the issue of atomFamily. You use `useAtom` generics for read-only atoms.

It's better to infer types:
```diff
-   const [state, setState] = useAtom<ComponentState<ExampleComponentState>>(
+   const [state, setState] = useAtom(
      jotaiComponentStateAtomFamily(id)
    );
```

**Reply by oferitz** [reply]
OK, understood, but I'm still grappling with how to specify the generic type for the initializeAtom. Specifically, I'd like to designate `ExampleComponentState` as the type `T`. What I'm trying to convey is that if I allow TypeScript to infer the type, it results in an extensive union type encompassing all possible types that the data property might have.

My preference is to be explicit and instruct the atom to use a designated type for the data field during its creation. In other words, I want to clearly communicate that when creating this atom, the specified type for the data field should be `ExampleComponentState`. I hope this clarifies my point.

**Reply by dai-shi** [reply]
If I understand your requirement correctly, it's not currently possible with TypeScript.
I think we have a similar discussion before. There's nothing a library can do.

If Recoil works in some ways, it should be asserting types.

Here's an example with type assertion: https://tsplay.dev/WPOjJm
I don't think it's something you are looking for.

If I misunderstand something, it would be helpful if you can create a smaller example with the typescript playground.

**Reply by oferitz** [reply]
Thanks for the answer.
This is how the type of atomFamily looks like in Recoil:
```ts
 export function atomFamily<T, P extends SerializableParam>(
  options: AtomFamilyOptions<T, P>,
 ): (param: P) => RecoilState<T>;
 ```
 as you can see it takes `T` and `P`  and you can see that `T` is the atom value and this allow me to pass the generic that i want to this atom upon creation:
 ```ts
 // atomFamily definition
 export const recoilComponentStateAtomFamily = recoilAtomFamily<
  ComponentState,
  string
>({
  key: "recoilComponentStateAtomFamily",
  default: (id: string) => {
    return getDefaultValue(id);
  },
});
// atom family usage
  const [state, setState] = useRecoilState<
    ComponentState<ExampleComponentState>
  >(recoilComponentStateAtomFamily(id));
  ```
  You can see the full example code of both Recoil and Jotai [HERE](https://codesandbox.io/p/sandbox/recoil-vs-jotai-atomfamily-ts-yxscv3?file=%2Fsrc%2FRecoilComponent.tsx%3A2%2C55&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpztlz8t00063b6g45vomp1i%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpztlz8t00023b6gtelsn5m0%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpztlz8t00033b6g0zg6yjrt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpztlz8t00053b6gz7y9mv7a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B51.868647240631155%252C48.131352759368845%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpztlz8t00023b6gtelsn5m0%2522%253A%257B%2522id%2522%253A%2522clpztlz8t00023b6gtelsn5m0%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq0wi58w00bj3b6ghghiorgl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A1%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A122%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clq22hsq000023b6hs45snxuu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A2%252C%2522startColumn%2522%253A55%252C%2522endLineNumber%2522%253A2%252C%2522endColumn%2522%253A55%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FRecoilComponent.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clq22hsq000023b6hs45snxuu%2522%257D%252C%2522clpztlz8t00053b6gz7y9mv7a%2522%253A%257B%2522id%2522%253A%2522clpztlz8t00053b6gz7y9mv7a%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpztlz8t00043b6g0jvp1x9i%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpztlz8t00043b6g0jvp1x9i%2522%257D%252C%2522clpztlz8t00033b6g0zg6yjrt%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpztlz8t00033b6g0zg6yjrt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
  It seems that your provided example with type assertion could work but obviously it is more verbose and type loose than Recoil's way. 
  Can you think of any way the same could work with Jotai?

**Reply by dai-shi** [reply, marked as answer]
https://tsplay.dev/mxz6zw

Okay, so I don't think I misunderstand your question.
Reocoil's atomFamily also returns `any`. So, it's the same. Internally, useRecoilState should do similar to type assertion. It's probably type annotation, so it's technically different, but atomFamily type is `any` anyways.

So, it's not type loose at all. If it's verbose, you can create your own hook to hide the type assertion. Let me know if you need help creating such a hook.

**Reply by oferitz** [reply]
I understand your perspective now, and indeed, it makes sense to encapsulate the verbosity within a custom hook. Thank you for your assistance; I appreciate it.

**Comment by oferitz** [comment]
This could be useful for others who run into a similar scenario. Based on @dai-shi advice and suggestions, I've crafted a solution that aligns with my requirements.

```ts
// atomFamily creation
export const jotaiComponentStateAtomFamily = atomFamily<
  string,
  PrimitiveAtom<ComponentState>
>((id) => atom(getDefaultValue(id)), deepEqual);

// The custom hook
export const useAtomState = <T>(id: string) => {
  return useAtom<PrimitiveAtom<ComponentState<T>>>(
    jotaiComponentStateAtomFamily(id)
  );
};

 // use in component
 const [state, setState] = useAtomState<ExampleComponentState>(id);
 ```
